src/main/scala/org/allenai/plugins/DeployPlugin.scala

```diff
@@ -236,64 +300,79 @@ object DeployPlugin extends AutoPlugin {
 
-    val deploys: Seq[Future[Try[Unit]]] = deployConfig.hostConfigs map { hostConfig =>
-      Future {
-        // Command to pass to rsync's "rsh" flag, and to use as the base of our ssh operations.
-        val sshCommand = {
-          val keyFile = sys.env.getOrElse("AWS_PEM_FILE", {
-            throw new IllegalStateException(s"Environment variable 'AWS_PEM_FILE' is undefined")
-          })
-          require(keyFile.nonEmpty, "Environment variable 'AWS_PEM_FILE' is empty")
-
-          Seq("ssh", "-i", keyFile, "-l", hostConfig.sshUser)
-        }
-        val deployHost = hostConfig.host
-        val deployDirectory = hostConfig.directory
-
-        val rsyncDirs = deployDirs.value map (name => s"--include=/$name")
-        val rsyncCommand = Seq.concat(
-          Seq("rsync", "-vcrtzP", "--rsh=" + sshCommand.mkString(" ")),
-          rsyncDirs,
-          Seq(
-            "--exclude=/*",
-            "--delete",
-            universalStagingDir.getPath + "/",
-            deployHost + ":" + deployDirectory
-          )
-        )
-
-        // Shell-friendly version of rsync command, with rsh value quoted.
-        val quotedRsync = rsyncCommand.patch(
-          2, Seq("--rsh=" + sshCommand.mkString("\"", " ", "\"")), 1
-        ).mkString(" ")
-
-        // Now, ssh to the remote host and run the restart script.
-        val restartScript = deployDirectory + "/" + hostConfig.startupScript
-        val restartArgs = if (hostConfig.startupArgs.nonEmpty) {
-          "--" +: hostConfig.startupArgs
-        } else {
-          Seq()
-        }
+    // Command to pass to rsync's "rsh" flag, and to use as the base of our ssh operations.
+    val sshCommand = {
+      val keyFile = sys.env.getOrElse("AWS_PEM_FILE", {
+        throw new IllegalStateException(s"Environment variable 'AWS_PEM_FILE' is undefined")
+      })
+      require(keyFile.nonEmpty, "Environment variable 'AWS_PEM_FILE' is empty")
 
-        val restartCommand = Seq.concat(
-          sshCommand,
-          Seq(deployHost, restartScript, "restart"),
-          restartArgs
-        )
+      Seq("ssh", "-i", keyFile, "-l", deployConfig.sshUser)
+    }
 
-        // Run commands within a Try so we can collect errors using Future.fold.
-        // TODO(danm): Use ProcessIO to capture and report the actual stderr of failures here.
-        Try {
-          log.info("Running " + quotedRsync + " . . .")
-          if (Process(rsyncCommand).! != 0) {
-            sys.error(s"Error running rsync to host '$deployHost'.")
+    // Loop over all unique remote hosts listed in config, deploying to them in parallel.
+    val hosts = deployConfig.replicaOverridesByHost.keys
+    val deploys: Iterable[Future[Try[Unit]]] = hosts flatMap { hostName =>
+      val baseDeployDir = deployConfig.baseDirectory
+
+      val replicas = deployConfig.replicaOverridesByHost(hostName)
+      val numReplicas = replicas.length
+
+      // First, ensure parent directory structure of deploy exists.
```

Comment:
Should I still attempt to run the `mkdir` anyways? If permission is denied, the error should pop out.